075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,490 @@
+The scaling operation ends during the same reconciliation, because after having the `StrimziPodSet` controller asking for new pods through the Kubernetes API, the subsequent calls to `podsReady()`, `serviceEndpointReady()` and `headlessServiceEndpointReady()` are waiting for the newly added pods, together with the corresponding services, to be ready.
+It means that the auto-rebalancing can start in the same reconciliation, being sure that the newly added pods are available for the upcoming rebalancing operation.
+Of course, the rebalancing will take time going through several reconciliations.
+
+For the scaling down, if the brokers to be removed are hosting topics' partitions, the operation is reverted before calling the `KafkaReconciler.reconcile()` method which is called with the original replicas count as if the cluster was never scaled down.
+If the brokers to be removed are not hosting any topics' partitions, the `KafkaReconciler.reconcile()` method will be called with the new replicas count and the scaling down is triggered by the `scaleDown()` call which updates the involved `StrimziPodSet`(s) so that the corresponding controller will take care of removing the brokers.
+In this case, the auto-rebalancing can start in the same reconciliation to take topics' partitions off but going through several reconciliations.
+When the auto-rebalancing ends, the triggered pods deletion takes some time but it doesn't have any impact on the rebalancing already done.
+Of course, if the scale down check doesn't fail, it means that no topics' partitions are hosted on the brokers to be removed and no auto-rebalancing is needed.
+
+The idea is about having a new `KafkaAutoRebalancingReconciler` class with its own `reconcile()` method which is called by the `KafkaAssemblyOperator` as one of the last steps in the reconciliation loop, so that:
+
+* we can get all the information about the brokers removed or added, which are available through the `KafkaClusterCreator` class (read later for more details).
+* we are sure that, in case of scaling up, the Cruise Control pod was restarted and it's now up and running again so it can accept new rebalancing requests.
+Of course the `KafkaAutoRebalancingReconciler.reconcile()` method also checks the status of an ongoing rebalancing across several reconciliations.
+
+It's also important taking into account that a user, by using `KafkaNodePool`(s), could remove and add brokers at the same time, so actually triggering both scaling down and up.
+In terms of scaling, the reconcile loop has the scale down triggered before the scaling up but it could be skipped because the brokers to be removed are still hosting topics' partitions. In such case the scaling up runs before.
+In terms of rebalancing, after the scaling up happens, the rebalancing for brokers to be removed takes the precedence on the rebalancing for newly added brokers. They are executed sequentially and not in parallel.
```

Comment:
The other way around will take more reconciliations (so more time). Let's take the following as an example ...
Cluster with brokers 0,1,2,3,4,5 ... user wants to remove 2,3 and add 6,7.
Doing the current approach (rebalance scale down first, rebalance scale up later):

- rebalance scale down, move partitions off brokers 2,3
- 2,3 removed (because we are sure they are empty), scales up 6,7 and rebalance scale up to 6,7

Doing the other way around:

- scale up 6,7 rebalance scale up to 6,7 (some partitions from 2,3 could be moved to 6,7 but 2,3 could still host others)
- rebalance scale down, move remaining partitions off brokers 2,3
- 2, 3 removed

Each step is one reconciliation or a group of reconciliations, depending on how long is the rebalancing. Anyway it shows the opposite order will take longer.

Reply:
You never know how long a rebalance would take, as there may be other tasks in the CC queue scheduled before that, but my point was about HA. Doing the scale down first, you run the cluster with less nodes, and a single node crash may cause under replicated partitions. This risk is much lower if you run the scale up rebalancing first.

Reply:
But makes everything more complex imho. I would not change the algorithm for now. It's already complex by itself :-)

Reply:
@fvaleri As far as I understood, it does not change the replication factor or anything. So I do not think it impacts HA. It will still have the same exposure to node failure as before. It also seems to me as if you are mixing two different events -> the scale-up (i.e. adding new nodes) and the scale-up rebalance (i.e. doing the rebalance for the empty nodes). 

If you for example scale one node pool from 6 nodes to 3 and in the same reconciliation add another node-pool with 3 nodes what will happen is the following:
* The scale-down will be initially reverted due to non-empty nodes
* The scale-up (new node pool) will be done and the new nodes will be added
* The auto-rebalancing as proposed would trigger the remove-nodes rebalance with the new nodes being there.
* => Not sure Cruise Control is smart enough to say "I'm removing these 3 nodes and here are my 3 brand new completely empty nodes, let's just move the data between them". But I guess some part of the data from the nodes that will be removed will go to the new nodes anyway regardless what Paolo does first.
 
I think that in the best case, if CC is smart enough, Paolos's approach does both with one rebalance and is better. In the worst case, if CC is not that smart, the amount of data movement would and time spent in summary should be IMHO roughly the same regardless of what goes first.

In reality, it is likely the scale-up and scale-down will be in most cases separate events (because you change two different node pools which generates two separate events) and the priority will not matter much and it will depend on which part you did first.

(@ppatierno That said, what happens if I have topics with RF=3 and try to scale down to 2 nodes? Does CC give us some reasonable error handling?)

Reply:
> That said, what happens if I have topics with RF=3 and try to scale down to 2 nodes? Does CC give us some reasonable error handling?

I guess it would depend on the goals you set? With the default ones having `RackAwareGoal` as hard goal by default I get:

```shell
Error processing POST request '/remove_broker' due to: 'com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException: [RackAwareGoal] Insufficient number of racks to distribute each replica (Current: 2, Needed: 3). Add at least 1 rack with brokers. Add at least 1 rack with brokers.'.
java.util.concurrent.ExecutionException: com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException: [RackAwareGoal] Insufficient number of racks to distribute each replica (Current: 2, Needed: 3). Add at least 1 rack with brokers. Add at least 1 rack with brokers.
```

(I just tested on my minikube instance).

Reply:
Well, no rebalance should work for your on Minikube with RackAware goal, or?

Reply:
Yeah in this case you are right but even not having the rackaware goal it says cannot move replica `Add at least 1 broker` so we are covered.

Reply:
Great, thanks.